src/StorageFlattening.cpp

```diff
@@ -206,18 +215,69 @@ class FlattenDimensions : public IRMutator {
             }
         }
 
-        Expr idx = mutate(flatten_args(op->name, op->args, Buffer<>(), output_buf));
         Expr value = mutate(op->values[0]);
-        stmt = Store::make(op->name, value, idx, output_buf, const_true(value.type().lanes()));
+        if (in_shader && !shader_scope_realizations.contains(op->name)) {
+            user_assert(op->args.size() == 3)
+                << "Image stores require three coordinates.\n";
+            Expr buffer_var =
+                Variable::make(type_of<halide_buffer_t *>(), op->name + ".buffer", output_buf);
+            vector<Expr> args = {
+                op->name, buffer_var,
+                op->args[0], op->args[1], op->args[2],
+                value};
+            Expr store = Call::make(value.type(), Call::image_store,
+                                    args, Call::Intrinsic);
+            stmt = Evaluate::make(store);
+        } else {
+            Expr idx = mutate(flatten_args(op->name, op->args, Buffer<>(), output_buf));
+            stmt = Store::make(op->name, value, idx, output_buf, const_true(value.type().lanes()));
+        }
     }
 
     void visit(const Call *op) {
         if (op->call_type == Call::Halide ||
             op->call_type == Call::Image) {
+
             internal_assert(op->value_index == 0);
-            Expr idx = mutate(flatten_args(op->name, op->args, op->image, op->param));
-            expr = Load::make(op->type, op->name, idx, op->image, op->param,
-                              const_true(op->type.lanes()));
+
+            if (in_shader && !shader_scope_realizations.contains(op->name)) {
+                ReductionDomain rdom;
+                Expr buffer_var =
+                    Variable::make(type_of<halide_buffer_t *>(), op->name + ".buffer",
+                                   op->image, op->param, rdom);
+
+                // Create image_load("name", name.buffer, x - x_min, x_extent,
+                // y - y_min, y_extent, ...).  Extents can be used by
+                // successive passes. OpenGL, for example, uses them
+                // for coordinate normalization.
+                vector<Expr> args(2);
+                args[0] = op->name;
+                args[1] = buffer_var;
+                for (size_t i = 0; i < op->args.size(); i++) {
+                    Expr min = make_shape_var(op->name, "min", i, op->image, op->param);
+                    Expr extent = make_shape_var(op->name, "extent", i, op->image, op->param);
+                    args.push_back(mutate(op->args[i]) - min);
+                    args.push_back(extent);
+                }
+                for (size_t i = op->args.size(); i < 3; i++) {
```

Comment:
Not sure I agree. Call nodes have arbitrary arity. image_load must have 3 args (currently). If we relax that we'd have to deal with at every site that analyzes an image_load. Easier to enforce it by construction.